File: agent-service/app/tools/med_schedule.py
"""
CareTopicz Agent Service - Regulated medication schedule tool.

Manage iPLEDGE, biologic, and REMS medication schedules.
"""

from typing import Any, Literal

from pydantic import BaseModel, Field

from app.clients.openemr import (
    cancel_medication_schedule,
    check_schedule_conflicts,
    complete_milestone,
    complete_treatment,
    create_medication_schedule,
    discontinue_medication_schedule,
    extend_medication_schedule,
    get_dashboard_alerts,
    get_medication_schedule,
    get_patient_demographics,
    pause_medication_schedule,
    reschedule_milestone,
    resume_medication_schedule,
)


class MedicationScheduleInput(BaseModel):
    """Input schema for medication_schedule."""

    action: Literal[
        "create",
        "status",
        "complete",
        "cancel",
        "reschedule",
        "conflicts",
        "dashboard",
        "extend",
        "complete_treatment",
        "discontinue",
        "pause",
        "resume",
    ] = Field(
        ...,
        description="Action: create, status, complete, cancel, reschedule, conflicts, dashboard, extend, complete_treatment, discontinue, pause, resume",
    )
    patient_id: int | None = Field(
        default=None,
        description="OpenEMR patient ID (required for create, status)",
    )
    medication: str | None = Field(
        default=None,
        description="Medication name (required for create), e.g. isotretinoin, adalimumab",
    )
    patient_category: Literal["fcbp", "non_fcbp_female", "male", "all"] | None = Field(
        default=None,
        description="fcbp, non_fcbp_female, or male (for create)",
    )
    milestone_id: int | None = Field(
        default=None,
        description="Milestone ID (for complete, reschedule)",
    )
    completed_date: str | None = Field(
        default=None,
        description="Date completed YYYY-MM-DD (for complete)",
    )
    new_due_date: str | None = Field(
        default=None,
        description="New due date YYYY-MM-DD (for reschedule)",
    )
    schedule_id: int | None = Field(
        default=None,
        description="Schedule ID (for cancel, conflicts)",
    )
    reason: str | None = Field(
        default=None,
        description="Reason for cancellation (for cancel)",
    )
    notes: str | None = Field(
        default=None,
        description="Additional notes",
    )
    start_date: str | None = Field(
        default=None,
        description="Start date YYYY-MM-DD (for create)",
    )
    created_by: str | None = Field(
        default="agent",
        description="Username of staff creating (for create)",
    )
    duration_months: int | None = Field(
        default=None,
        description="Number of months of milestones to generate (default: 6 for isotretinoin, 3 for biologics). For create or extend.",
    )


def _patient_name(patient_id: int) -> str:
    """Get patient display name."""
    demos = get_patient_demographics(patient_id)
    if not demos:
        return f"Patient {patient_id}"
    f = demos.get("fname", "")
    l = demos.get("lname", "")
    return f"{f} {l}".strip() or f"Patient {patient_id}"


def _format_schedule_status(data: dict[str, Any], patient_id: int | None = None) -> str:
    """Format schedule/milestone data for LLM."""
    if not data.get("success"):
        return f"Error: {data.get('error', 'Unknown error')}"

    sched = data.get("schedule")
    schedules = data.get("schedules", [])

    if sched:
        lines = []
        pid = sched.get("patient_id")
        name = _patient_name(int(pid)) if pid else (_patient_name(patient_id) if patient_id else None)
        if name:
            lines.append(f"Patient: {name}")
        lines.append(f"Schedule ID: {sched.get('id')} | Status: {sched.get('status', '')}")
        lines.append(f"Protocol: {sched.get('patient_category', '')} | Start: {sched.get('start_date', '')}")
        milestones = sched.get("milestones", [])
        pending = [m for m in milestones if m.get("status") in ("pending", "scheduled", "overdue")]
        if pending:
            lines.append("\nNext pending:")
            for m in pending[:5]:
                due = m.get("due_date", "")
                wend = m.get("window_end", "")
                status = m.get("status", "pending")
                mid = m.get('id', '')
            lines.append(f"  - [milestone_id={mid}] {m.get('step_name', '')}: due {due} (window ends {wend}) [{status}]")
        else:
            lines.append("\nAll milestones completed or no milestones.")
        if data.get("warnings"):
            lines.append("\nWarnings: " + "; ".join(data["warnings"]))
        return "\n".join(lines)

    if schedules is not None:
        if not schedules:
            return "No active schedules found."
        lines = []
        for s in schedules:
            pid = s.get("patient_id")
            name = _patient_name(int(pid)) if pid else "Unknown"
            mname = s.get("medication_name", "")
            lines.append(f"- {name}: Schedule {s.get('id')} | {mname} | Status: {s.get('status', '')}")
            milestones = s.get("milestones", [])
            pending = [m for m in milestones if m.get("status") in ("pending", "scheduled", "overdue")]
            if pending:
                m0 = pending[0]
                lines.append(f"  Next: {m0.get('step_name', '')} due {m0.get('due_date', '')}")
        return "Active schedules:\n" + "\n".join(lines)

    return "No schedule data returned."


def medication_schedule(
    action: str,
    patient_id: int | None = None,
    medication: str | None = None,
    patient_category: str | None = None,
    milestone_id: int | None = None,
    completed_date: str | None = None,
    new_due_date: str | None = None,
    schedule_id: int | None = None,
    reason: str | None = None,
    notes: str | None = None,
    start_date: str | None = None,
    created_by: str = "agent",
    duration_months: int | None = None,
) -> dict[str, Any]:
    """
    Manage regulated medication schedules (iPLEDGE, biologics, REMS).
    """
    if action == "create":
        if not patient_id or not medication or not patient_category:
            return {"success": False, "error": "patient_id, medication, and patient_category required for create"}
        from datetime import date
        sd = start_date or str(date.today())
        result = create_medication_schedule(
            patient_id=patient_id,
            medication=medication,
            patient_category=patient_category,
            start_date=sd,
            created_by=created_by or "agent",
            duration_months=duration_months,
            notes=notes or "",
        )
        if result.get("success") and result.get("schedule"):
            return result
        return result

    if action == "extend":
        if not schedule_id and not patient_id:
            return {"success": False, "error": "schedule_id or patient_id required for extend"}
        if not duration_months:
            return {"success": False, "error": "duration_months required for extend (how many months to add)"}
        result = extend_medication_schedule(
            schedule_id=schedule_id,
            patient_id=patient_id,
            duration_months=duration_months or 3,
        )
        return result

    if action == "complete_treatment":
        if not schedule_id and not patient_id:
            return {"success": False, "error": "schedule_id or patient_id required for complete_treatment"}
        result = complete_treatment(
            schedule_id=schedule_id,
            patient_id=patient_id,
            notes=notes or "",
        )
        return result

    if action == "discontinue":
        if not schedule_id and not patient_id:
            return {"success": False, "error": "schedule_id or patient_id required for discontinue"}
        result = discontinue_medication_schedule(
            schedule_id=schedule_id,
            patient_id=patient_id,
            reason=reason or "Discontinued",
            discontinued_by=created_by or "agent",
        )
        if result.get("success"):
            return {"success": True, "message": f"Schedule {schedule_id or 'for patient ' + str(patient_id)} has been successfully discontinued. Reason: {reason or 'Discontinued'}. No further action needed."}
        return result

    if action == "pause":
        if not schedule_id and not patient_id:
            return {"success": False, "error": "schedule_id or patient_id required for pause"}
        result = pause_medication_schedule(
            schedule_id=schedule_id,
            patient_id=patient_id,
            notes=notes or "",
        )
        return result

    if action == "resume":
        if not schedule_id and not patient_id:
            return {"success": False, "error": "schedule_id or patient_id required for resume"}
        result = resume_medication_schedule(
            schedule_id=schedule_id,
            patient_id=patient_id,
        )
        return result

    if action == "status":
        if not patient_id:
            return {"success": False, "error": "patient_id required for status"}
        result = get_medication_schedule(patient_id)
        return result

    if action == "complete":
        if not milestone_id:
            return {"success": False, "error": "milestone_id required for complete"}
        from datetime import date
        cd = completed_date or str(date.today())
        result = complete_milestone(
            milestone_id=milestone_id,
            completed_by=created_by or "agent",
            completed_date=cd,
            notes=notes or "",
        )
        if result.get("success") and result.get("schedule"):
            pid = result["schedule"].get("patient_id")
            result["_formatted"] = _format_schedule_status(result, int(pid) if pid else None)
        return result

    if action == "cancel":
        if not schedule_id:
            return {"success": False, "error": "schedule_id required for cancel"}
        result = cancel_medication_schedule(
            schedule_id=schedule_id,
            reason=reason or "Cancelled",
            cancelled_by=created_by or "agent",
        )
        if result.get("success"):
            return {"success": True, "message": f"Schedule {schedule_id} has been successfully cancelled. Reason: {reason or 'Cancelled'}. No further action needed."}
        return result

    if action == "reschedule":
        if not milestone_id or not new_due_date:
            return {"success": False, "error": "milestone_id and new_due_date required for reschedule"}
        return reschedule_milestone(
            milestone_id=milestone_id,
            new_due_date=new_due_date,
            rescheduled_by=created_by or "agent",
        )

    if action == "conflicts":
        if not schedule_id:
            return {"success": False, "error": "schedule_id required for conflicts"}
        return check_schedule_conflicts(schedule_id)

    if action == "dashboard":
        return get_dashboard_alerts()

    return {"success": False, "error": f"Unknown action: {action}"}

File: interface/modules/custom_modules/mod-ai-agent/public/med_schedule.php
<?php
/**
 * CareTopicz Bounty - Regulated Medication Schedule CRUD
 * Internal-only endpoint for agent and clinic staff.
 */

$remoteIp = $_SERVER['REMOTE_ADDR'] ?? '';
$isInternal = in_array($remoteIp, ['127.0.0.1', '::1'])
    || strpos($remoteIp, '172.') === 0
    || strpos($remoteIp, '10.') === 0;

if (!$isInternal) {
    http_response_code(403);
    header('Content-Type: application/json');
    echo json_encode(['error' => 'Forbidden - internal access only']);
    exit;
}

header('Content-Type: application/json');

$siteDir = '/var/www/localhost/htdocs/openemr/sites/default';
if (isset($GLOBALS['OE_SITE_DIR'])) {
    $siteDir = $GLOBALS['OE_SITE_DIR'];
}
$sqlConfFile = $siteDir . '/sqlconf.php';

if (!file_exists($sqlConfFile)) {
    echo json_encode(['success' => false, 'error' => 'Database config not found']);
    exit;
}

include $sqlConfFile;

try {
    $dsn = "mysql:host={$host};dbname={$dbase};charset=utf8mb4";
    $pdo = new PDO($dsn, $login, $pass, [
        PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION,
        PDO::ATTR_DEFAULT_FETCH_MODE => PDO::FETCH_ASSOC,
    ]);
} catch (PDOException $e) {
    echo json_encode(['success' => false, 'error' => 'Database connection failed']);
    exit;
}

$action = $_GET['action'] ?? $_POST['action'] ?? '';

// Parse JSON body for POST
$body = [];
if ($_SERVER['REQUEST_METHOD'] === 'POST' && ($raw = file_get_contents('php://input'))) {
    $body = json_decode($raw, true) ?: [];
}

function jsonResp($data) {
    echo json_encode($data);
}

function getProtocol($pdo, $medication, $patientCategory) {
    $cat = $patientCategory;
    if ($patientCategory === 'all') {
        $cat = 'all';
    }
    $stmt = $pdo->prepare("
        SELECT * FROM medication_protocols
        WHERE medication_name = ? AND patient_category = ?
    ");
    $stmt->execute([$medication, $cat]);
    $row = $stmt->fetch();
    if (!$row) {
        $stmt = $pdo->prepare("
            SELECT * FROM medication_protocols
            WHERE medication_name = ? AND (patient_category = ? OR patient_category = 'all')
            ORDER BY patient_category = 'all' ASC
            LIMIT 1
        ");
        $stmt->execute([$medication, $cat]);
        $row = $stmt->fetch();
    }
    return $row;
}

function calcDueDates($steps, $startDate, $prevCompletedDate = null) {
    $dueDates = [];
    $prev = $prevCompletedDate ? strtotime($prevCompletedDate) : strtotime($startDate);
    foreach ($steps as $i => $step) {
        $step = is_string($step) ? json_decode($step, true) : $step;
        if (!$step) continue;
        $daysFromPrev = $step['days_from_prev'] ?? $step['days_from_start'] ?? 0;
        $windowDays = $step['window_days'] ?? 7;
        $due = strtotime("+{$daysFromPrev} days", $prev);
        $dueDate = date('Y-m-d', $due);
        $windowEnd = date('Y-m-d', strtotime("+{$windowDays} days", $due));
        $dueDates[] = [
            'step_number' => $i + 1,
            'step_name' => $step['name'],
            'step_type' => $step['type'],
            'description' => $step['description'] ?? null,
            'due_date' => $dueDate,
            'window_start' => $dueDate,
            'window_end' => $windowEnd,
        ];
        $prev = $due;
    }
    return $dueDates;
}

switch ($action) {
    case 'get_protocols':
        $medication = $_GET['medication'] ?? '';
        $sql = "SELECT id, medication_name, protocol_type, patient_category, source FROM medication_protocols";
        $params = [];
        if ($medication) {
            $sql .= " WHERE medication_name = ?";
            $params[] = $medication;
        }
        $sql .= " ORDER BY medication_name, patient_category";
        $stmt = $pdo->prepare($sql);
        $stmt->execute($params);
        $rows = $stmt->fetchAll();
        jsonResp(['success' => true, 'protocols' => $rows]);
        break;

    case 'get_schedule':
        $patientId = (int)($_GET['patient_id'] ?? 0);
        if (!$patientId) {
            jsonResp(['success' => false, 'error' => 'patient_id required']);
            break;
        }
        $stmt = $pdo->prepare("
            SELECT s.*, p.medication_name, p.protocol_type, p.patient_category
            FROM patient_med_schedules s
            JOIN medication_protocols p ON p.id = s.protocol_id
            WHERE s.patient_id = ? AND s.status NOT IN ('completed', 'cancelled')
        ");
        $stmt->execute([$patientId]);
        $schedules = $stmt->fetchAll();
        $milestones = [];
        foreach ($schedules as &$sch) {
            $mStmt = $pdo->prepare("SELECT * FROM schedule_milestones WHERE schedule_id = ? ORDER BY step_number");
            $mStmt->execute([$sch['id']]);
            $sch['milestones'] = $mStmt->fetchAll();
            $milestones = array_merge($milestones, $sch['milestones']);
        }
        jsonResp(['success' => true, 'schedules' => $schedules]);
        break;

    case 'get_dashboard_alerts':
        $stmt = $pdo->prepare("
            SELECT m.*, s.patient_id, s.status as schedule_status, p.medication_name, p.protocol_type
            FROM schedule_milestones m
            JOIN patient_med_schedules s ON s.id = m.schedule_id
            JOIN medication_protocols p ON p.id = s.protocol_id
            WHERE m.status IN ('pending', 'scheduled', 'overdue')
            AND s.status IN ('initiating', 'active', 'completing')
            AND (m.due_date <= DATE_ADD(CURDATE(), INTERVAL 7 DAY) OR m.window_end <= DATE_ADD(CURDATE(), INTERVAL 3 DAY))
            ORDER BY m.due_date
        ");
        $stmt->execute();
        jsonResp(['success' => true, 'alerts' => $stmt->fetchAll()]);
        break;

    case 'get_all_active':
        $stmt = $pdo->prepare("
            SELECT s.*, p.medication_name, p.protocol_type,
                   (SELECT COUNT(*) FROM schedule_milestones WHERE schedule_id = s.id AND status IN ('pending','overdue')) as pending_count
            FROM patient_med_schedules s
            JOIN medication_protocols p ON p.id = s.protocol_id
            WHERE s.status IN ('initiating', 'active', 'completing')
        ");
        $stmt->execute();
        jsonResp(['success' => true, 'schedules' => $stmt->fetchAll()]);
        break;

    case 'create_schedule':
        $patientId = (int)($body['patient_id'] ?? 0);
        $medication = trim($body['medication'] ?? '');
        $patientCategory = $body['patient_category'] ?? 'male';
        $createdBy = $body['created_by'] ?? 'agent';
        $startDate = $body['start_date'] ?? date('Y-m-d');

        if (!$patientId || !$medication) {
            jsonResp(['success' => false, 'error' => 'patient_id and medication required']);
            break;
        }

        $protocol = getProtocol($pdo, $medication, $patientCategory);
        if (!$protocol) {
            jsonResp(['success' => false, 'error' => "No protocol found for {$medication} / {$patientCategory}"]);
            break;
        }

        $existStmt = $pdo->prepare("
            SELECT id FROM patient_med_schedules
            WHERE patient_id = ? AND protocol_id = ? AND status IN ('initiating','active','completing')
        ");
        $existStmt->execute([$patientId, $protocol['id']]);
        if ($existStmt->fetch()) {
            jsonResp(['success' => false, 'error' => 'Active schedule already exists for this patient and medication']);
            break;
        }

        $pdo->beginTransaction();
        try {
            $ins = $pdo->prepare("
                INSERT INTO patient_med_schedules (patient_id, protocol_id, patient_category, status, created_by, start_date)
                VALUES (?, ?, ?, 'initiating', ?, ?)
            ");
            $ins->execute([$patientId, $protocol['id'], $patientCategory, $createdBy, $startDate]);
            $scheduleId = $pdo->lastInsertId();

            $steps = json_decode($protocol['steps'], true);
            if ($steps) {
                $milestoneIns = $pdo->prepare("
                    INSERT INTO schedule_milestones (schedule_id, step_number, step_name, step_type, description, due_date, window_start, window_end)
                    VALUES (?, ?, ?, ?, ?, ?, ?, ?)
                ");
                $prevDate = $startDate;
                foreach ($steps as $i => $step) {
                    $daysFromPrev = $step['days_from_prev'] ?? $step['days_from_start'] ?? 0;
                    $windowDays = $step['window_days'] ?? 7;
                    $due = date('Y-m-d', strtotime("+{$daysFromPrev} days", strtotime($prevDate)));
                    $windowEnd = date('Y-m-d', strtotime("+{$windowDays} days", strtotime($due)));
                    $milestoneIns->execute([
                        $scheduleId,
                        $i + 1,
                        $step['name'],
                        $step['type'],
                        $step['description'] ?? null,
                        $due,
                        $due,
                        $windowEnd,
                    ]);
                }
            }

            $monthlyCycle = json_decode($protocol['monthly_cycle'] ?? '{}', true);
            $durationMonths = $monthlyCycle['typical_duration_months'] ?? 6;
            if ($durationMonths && !empty($monthlyCycle['steps'])) {
                $cycleSteps = $monthlyCycle['steps'];
                $stepNum = count($steps) + 1;
                $firstPrescDate = $startDate;
                foreach ($steps as $st) {
                    if (($st['name'] ?? '') === 'first_prescription') {
                        $firstPrescDate = date('Y-m-d', strtotime($st['window_end'] ?? $startDate));
                        break;
                    }
                }
                for ($m = 1; $m <= $durationMonths; $m++) {
                    $monthStart = date('Y-m-d', strtotime("+{$m} months", strtotime($firstPrescDate)));
                    foreach ($cycleSteps as $cs) {
                        $due = $monthStart;
                        $windowEnd = date('Y-m-d', strtotime('+7 days', strtotime($due)));
                        $mIns = $pdo->prepare("
                            INSERT INTO schedule_milestones (schedule_id, step_number, step_name, step_type, description, due_date, window_start, window_end)
                            VALUES (?, ?, ?, ?, ?, ?, ?, ?)
                        ");
                        $mIns->execute([
                            $scheduleId,
                            $stepNum++,
                            $cs['name'] . '_m' . $m,
                            $cs['type'],
                            $cs['description'] ?? null,
                            $due,
                            $due,
                            $windowEnd,
                        ]);
                    }
                }
            }

            $pdo->commit();

            $schedStmt = $pdo->prepare("SELECT * FROM patient_med_schedules WHERE id = ?");
            $schedStmt->execute([$scheduleId]);
            $schedule = $schedStmt->fetch();
            $mStmt = $pdo->prepare("SELECT * FROM schedule_milestones WHERE schedule_id = ? ORDER BY step_number");
            $mStmt->execute([$scheduleId]);
            $schedule['milestones'] = $mStmt->fetchAll();

            jsonResp(['success' => true, 'schedule' => $schedule]);
        } catch (Exception $e) {
            $pdo->rollBack();
            jsonResp(['success' => false, 'error' => $e->getMessage()]);
        }
        break;

    case 'complete_milestone':
        $milestoneId = (int)($body['milestone_id'] ?? 0);
        $completedBy = $body['completed_by'] ?? 'agent';
        $completedDate = $body['completed_date'] ?? date('Y-m-d');
        $notes = $body['notes'] ?? '';

        if (!$milestoneId) {
            jsonResp(['success' => false, 'error' => 'milestone_id required']);
            break;
        }

        $mStmt = $pdo->prepare("SELECT * FROM schedule_milestones WHERE id = ?");
        $mStmt->execute([$milestoneId]);
        $milestone = $mStmt->fetch();
        if (!$milestone) {
            jsonResp(['success' => false, 'error' => 'Milestone not found']);
            break;
        }

        $upd = $pdo->prepare("
            UPDATE schedule_milestones SET status = 'completed', completed_date = ?, completed_by = ?, notes = ?
            WHERE id = ?
        ");
        $upd->execute([$completedDate, $completedBy, $notes, $milestoneId]);

        $schedStmt = $pdo->prepare("SELECT * FROM patient_med_schedules WHERE id = ?");
        $schedStmt->execute([$milestone['schedule_id']]);
        $schedule = $schedStmt->fetch();
        $mStmt = $pdo->prepare("SELECT * FROM schedule_milestones WHERE schedule_id = ? ORDER BY step_number");
        $mStmt->execute([$milestone['schedule_id']]);
        $schedule['milestones'] = $mStmt->fetchAll();

        jsonResp(['success' => true, 'schedule' => $schedule]);
        break;

    case 'cancel_schedule':
        $scheduleId = (int)($body['schedule_id'] ?? 0);
        $reason = $body['cancelled_reason'] ?? $body['reason'] ?? 'Cancelled';
        $cancelledBy = $body['cancelled_by'] ?? $body['completed_by'] ?? 'agent';

        if (!$scheduleId) {
            jsonResp(['success' => false, 'error' => 'schedule_id required']);
            break;
        }

        $pdo->prepare("UPDATE patient_med_schedules SET status = 'cancelled', cancelled_reason = ? WHERE id = ?")
            ->execute([$reason, $scheduleId]);
        $pdo->prepare("UPDATE schedule_milestones SET status = 'cancelled' WHERE schedule_id = ? AND status IN ('pending','scheduled','overdue')")
            ->execute([$scheduleId]);

        jsonResp(['success' => true, 'message' => 'Schedule cancelled']);
        break;

    case 'reschedule_milestone':
        $milestoneId = (int)($body['milestone_id'] ?? 0);
        $newDueDate = $body['new_due_date'] ?? '';
        $rescheduledBy = $body['rescheduled_by'] ?? 'agent';

        if (!$milestoneId || !$newDueDate) {
            jsonResp(['success' => false, 'error' => 'milestone_id and new_due_date required']);
            break;
        }

        $mStmt = $pdo->prepare("SELECT * FROM schedule_milestones WHERE id = ?");
        $mStmt->execute([$milestoneId]);
        $milestone = $mStmt->fetch();
        if (!$milestone) {
            jsonResp(['success' => false, 'error' => 'Milestone not found']);
            break;
        }

        $windowEnd = $milestone['window_end'];
        $warnings = [];
        if ($windowEnd && $newDueDate > $windowEnd) {
            $warnings[] = 'New date is outside the compliance window';
        }

        $upd = $pdo->prepare("UPDATE schedule_milestones SET due_date = ?, window_start = ? WHERE id = ?");
        $upd->execute([$newDueDate, $newDueDate, $milestoneId]);

        $schedStmt = $pdo->prepare("SELECT * FROM patient_med_schedules WHERE id = ?");
        $schedStmt->execute([$milestone['schedule_id']]);
        $schedule = $schedStmt->fetch();
        $mStmt = $pdo->prepare("SELECT * FROM schedule_milestones WHERE schedule_id = ? ORDER BY step_number");
        $mStmt->execute([$milestone['schedule_id']]);
        $schedule['milestones'] = $mStmt->fetchAll();

        jsonResp(['success' => true, 'schedule' => $schedule, 'warnings' => $warnings]);
        break;

    case 'check_conflicts':
        $scheduleId = (int)($_GET['schedule_id'] ?? 0);
        if (!$scheduleId) {
            jsonResp(['success' => false, 'error' => 'schedule_id required']);
            break;
        }

        $today = date('Y-m-d');
        $stmt = $pdo->prepare("
            SELECT * FROM schedule_milestones
            WHERE schedule_id = ? AND status IN ('pending', 'scheduled')
        ");
        $stmt->execute([$scheduleId]);
        $milestones = $stmt->fetchAll();
        $conflicts = [];

        foreach ($milestones as $m) {
            if ($m['due_date'] < $today) {
                $conflicts[] = ['milestone_id' => $m['id'], 'severity' => 'critical', 'message' => 'Overdue: ' . $m['step_name']];
            } elseif ($m['window_end'] && $m['window_end'] <= date('Y-m-d', strtotime('+3 days'))) {
                $conflicts[] = ['milestone_id' => $m['id'], 'severity' => 'warning', 'message' => 'Compliance window expiring: ' . $m['step_name']];
            }
        }

        jsonResp(['success' => true, 'conflicts' => $conflicts]);
        break;

    default:
        jsonResp(['success' => false, 'error' => 'Unknown action']);
}
